File: src/tools/finance/hmm.ts
/**
 * Gaussian Hidden Markov Model (HMM) for financial regime detection.
 *
 * Replaces threshold-based observable Markov chains with probabilistic
 * state inference. States are latent — the model learns emission distributions
 * (what returns look like in each state) and transition probabilities jointly
 * via the Baum-Welch (EM) algorithm.
 *
 * Advantages over observable Markov:
 * - No hard threshold artifacts (+1.01% "bull" vs +0.99% "sideways")
 * - Learns from data instead of fixed rules
 * - Provides probabilistic state beliefs, not hard assignments
 * - Naturally captures volatility clustering
 */

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface HMMParams {
  /** Number of hidden states */
  nStates: number;
  /** Initial state probabilities π[i] = P(state_0 = i) */
  pi: number[];
  /** Transition matrix A[i][j] = P(state_t+1 = j | state_t = i) */
  A: number[][];
  /** Emission means μ[i] for each state (Gaussian) */
  means: number[];
  /** Emission standard deviations σ[i] for each state (Gaussian) */
  stds: number[];
}

export interface HMMFitResult {
  params: HMMParams;
  logLikelihood: number;
  iterations: number;
  converged: boolean;
}

export interface HMMPrediction {
  /** Most likely current state (Viterbi) */
  currentState: number;
  /** Posterior probability of each state at each time step: gamma[t][i] */
  stateProbabilities: number[][];
  /** Posterior probability of each state at the LAST time step */
  currentStateProbabilities: number[];
  /** n-step ahead state probability forecast */
  forecastProbabilities: number[];
  /** Expected return (weighted by state probabilities and emission means) */
  expectedReturn: number;
  /** Expected volatility (weighted by state probabilities and emission stds) */
  expectedVolatility: number;
}

// ---------------------------------------------------------------------------
// Gaussian PDF
// ---------------------------------------------------------------------------

function gaussianPdf(x: number, mean: number, std: number): number {
  if (std < 1e-10) return x === mean ? 1e10 : 1e-300;
  const z = (x - mean) / std;
  return Math.exp(-0.5 * z * z) / (std * Math.sqrt(2 * Math.PI));
}

// ---------------------------------------------------------------------------
// Initialize HMM parameters using K-means-style heuristic
// ---------------------------------------------------------------------------

export function initializeHMM(observations: number[], nStates: number): HMMParams {
  const sorted = [...observations].sort((a, b) => a - b);
  const n = sorted.length;

  // Initialize means by quantile splitting
  const means: number[] = [];
  for (let i = 0; i < nStates; i++) {
    const idx = Math.floor((i + 0.5) * n / nStates);
    means.push(sorted[Math.min(idx, n - 1)]);
  }
  // Sort means so state 0 = bearish (lowest returns), state N-1 = bullish
  means.sort((a, b) => a - b);

  // Initialize stds as the std of each quantile segment
  const stds: number[] = [];
  for (let i = 0; i < nStates; i++) {
    const lo = Math.floor(i * n / nStates);
    const hi = Math.floor((i + 1) * n / nStates);
    const segment = sorted.slice(lo, hi);
    const segMean = segment.reduce((s, v) => s + v, 0) / segment.length;
    const segVar = segment.reduce((s, v) => s + (v - segMean) ** 2, 0) / segment.length;
    stds.push(Math.max(Math.sqrt(segVar), 1e-6));
  }

  // Uniform initial probabilities
  const pi = Array(nStates).fill(1 / nStates);

  // Slightly diagonal-dominant transition matrix (regime persistence)
  const A: number[][] = Array.from({ length: nStates }, (_, i) =>
    Array.from({ length: nStates }, (_, j) =>
      i === j ? 0.7 : 0.3 / (nStates - 1),
    ),
  );

  return { nStates, pi, A, means, stds };
}

// ---------------------------------------------------------------------------
// Forward algorithm: compute α[t][i] = P(obs_1..t, state_t = i)
// Uses log-scaling for numerical stability.
// ---------------------------------------------------------------------------

export function forward(obs: number[], params: HMMParams): {
  alpha: number[][];
  scales: number[];
  logLikelihood: number;
} {
  const { nStates, pi, A, means, stds } = params;
  const T = obs.length;
  const alpha: number[][] = Array.from({ length: T }, () => Array(nStates).fill(0));
  const scales: number[] = Array(T).fill(0);

  // t = 0
  for (let i = 0; i < nStates; i++) {
    alpha[0][i] = pi[i] * gaussianPdf(obs[0], means[i], stds[i]);
  }
  scales[0] = alpha[0].reduce((s, v) => s + v, 0);
  if (scales[0] < 1e-300) scales[0] = 1e-300;
  for (let i = 0; i < nStates; i++) alpha[0][i] /= scales[0];

  // t = 1..T-1
  for (let t = 1; t < T; t++) {
    for (let j = 0; j < nStates; j++) {
      let sum = 0;
      for (let i = 0; i < nStates; i++) {
        sum += alpha[t - 1][i] * A[i][j];
      }
      alpha[t][j] = sum * gaussianPdf(obs[t], means[j], stds[j]);
    }
    scales[t] = alpha[t].reduce((s, v) => s + v, 0);
    if (scales[t] < 1e-300) scales[t] = 1e-300;
    for (let j = 0; j < nStates; j++) alpha[t][j] /= scales[t];
  }

  const logLikelihood = scales.reduce((s, c) => s + Math.log(c), 0);
  return { alpha, scales, logLikelihood };
}

// ---------------------------------------------------------------------------
// Backward algorithm: compute β[t][i] = P(obs_t+1..T | state_t = i)
// ---------------------------------------------------------------------------

export function backward(obs: number[], params: HMMParams, scales: number[]): number[][] {
  const { nStates, A, means, stds } = params;
  const T = obs.length;
  const beta: number[][] = Array.from({ length: T }, () => Array(nStates).fill(0));

  // t = T-1
  for (let i = 0; i < nStates; i++) beta[T - 1][i] = 1;

  // t = T-2..0
  for (let t = T - 2; t >= 0; t--) {
    for (let i = 0; i < nStates; i++) {
      let sum = 0;
      for (let j = 0; j < nStates; j++) {
        sum += A[i][j] * gaussianPdf(obs[t + 1], means[j], stds[j]) * beta[t + 1][j];
      }
      beta[t][i] = sum / scales[t + 1];
    }
  }

  return beta;
}

// ---------------------------------------------------------------------------
// Baum-Welch (EM) algorithm: fit HMM parameters to observations
// ---------------------------------------------------------------------------

export function baumWelch(
  observations: number[],
  nStates = 3,
  maxIterations = 100,
  tolerance = 1e-4,
  minStd = 1e-4,
): HMMFitResult {
  const T = observations.length;
  if (T < 10) {
    // Not enough data — return default params
    const params = initializeHMM(observations, nStates);
    return { params, logLikelihood: -Infinity, iterations: 0, converged: false };
  }

  let params = initializeHMM(observations, nStates);
  let prevLL = -Infinity;
  let iterations = 0;
  let converged = false;

  for (let iter = 0; iter < maxIterations; iter++) {
    iterations = iter + 1;

    // E-step: forward-backward
    const { alpha, scales, logLikelihood } = forward(observations, params);
    const beta = backward(observations, params, scales);

    // Convergence check
    if (Math.abs(logLikelihood - prevLL) < tolerance) {
      converged = true;
      break;
    }
    prevLL = logLikelihood;

    // Compute gamma[t][i] = P(state_t = i | obs)
    const gamma: number[][] = Array.from({ length: T }, () => Array(nStates).fill(0));
    for (let t = 0; t < T; t++) {
      let sum = 0;
      for (let i = 0; i < nStates; i++) {
        gamma[t][i] = alpha[t][i] * beta[t][i];
        sum += gamma[t][i];
      }
      if (sum < 1e-300) sum = 1e-300;
      for (let i = 0; i < nStates; i++) gamma[t][i] /= sum;
    }

    // Compute xi[t][i][j] = P(state_t = i, state_t+1 = j | obs)
    const xi: number[][][] = Array.from({ length: T - 1 }, () =>
      Array.from({ length: nStates }, () => Array(nStates).fill(0)),
    );
    for (let t = 0; t < T - 1; t++) {
      let sum = 0;
      for (let i = 0; i < nStates; i++) {
        for (let j = 0; j < nStates; j++) {
          xi[t][i][j] =
            alpha[t][i] *
            params.A[i][j] *
            gaussianPdf(observations[t + 1], params.means[j], params.stds[j]) *
            beta[t + 1][j];
          sum += xi[t][i][j];
        }
      }
      if (sum < 1e-300) sum = 1e-300;
      for (let i = 0; i < nStates; i++) {
        for (let j = 0; j < nStates; j++) {
          xi[t][i][j] /= sum;
        }
      }
    }

    // M-step: update parameters
    const newPi = gamma[0].slice();
    const newA: number[][] = Array.from({ length: nStates }, () => Array(nStates).fill(0));
    const newMeans: number[] = Array(nStates).fill(0);
    const newStds: number[] = Array(nStates).fill(0);

    for (let i = 0; i < nStates; i++) {
      // Transition matrix
      let gammaSum = 0;
      for (let t = 0; t < T - 1; t++) gammaSum += gamma[t][i];
      if (gammaSum < 1e-300) gammaSum = 1e-300;

      for (let j = 0; j < nStates; j++) {
        let xiSum = 0;
        for (let t = 0; t < T - 1; t++) xiSum += xi[t][i][j];
        newA[i][j] = xiSum / gammaSum;
      }

      // Emission parameters
      let totalGamma = 0;
      for (let t = 0; t < T; t++) totalGamma += gamma[t][i];
      if (totalGamma < 1e-300) totalGamma = 1e-300;

      // Mean
      let weightedSum = 0;
      for (let t = 0; t < T; t++) weightedSum += gamma[t][i] * observations[t];
      newMeans[i] = weightedSum / totalGamma;

      // Std
      let weightedVarSum = 0;
      for (let t = 0; t < T; t++) {
        weightedVarSum += gamma[t][i] * (observations[t] - newMeans[i]) ** 2;
      }
      newStds[i] = Math.max(Math.sqrt(weightedVarSum / totalGamma), minStd);
    }

    // Ensure means are sorted (state 0 = bearish, state N-1 = bullish)
    const stateOrder = newMeans.map((m, i) => ({ m, i })).sort((a, b) => a.m - b.m);
    const sortedMeans = stateOrder.map(s => s.m);
    const sortedStds = stateOrder.map(s => newStds[s.i]);
    const sortedPi = stateOrder.map(s => newPi[s.i]);
    const sortedA: number[][] = Array.from({ length: nStates }, (_, i) =>
      Array.from({ length: nStates }, (_, j) => newA[stateOrder[i].i][stateOrder[j].i]),
    );

    params = {
      nStates,
      pi: sortedPi,
      A: sortedA,
      means: sortedMeans,
      stds: sortedStds,
    };
  }

  return { params, logLikelihood: prevLL, iterations, converged };
}

// ---------------------------------------------------------------------------
// Viterbi decoding: find most likely state sequence
// ---------------------------------------------------------------------------

export function viterbi(obs: number[], params: HMMParams): number[] {
  const { nStates, pi, A, means, stds } = params;
  const T = obs.length;

  // Use log probabilities for numerical stability
  const delta: number[][] = Array.from({ length: T }, () => Array(nStates).fill(-Infinity));
  const psi: number[][] = Array.from({ length: T }, () => Array(nStates).fill(0));

  // t = 0
  for (let i = 0; i < nStates; i++) {
    delta[0][i] = Math.log(Math.max(pi[i], 1e-300)) + Math.log(Math.max(gaussianPdf(obs[0], means[i], stds[i]), 1e-300));
  }

  // t = 1..T-1
  for (let t = 1; t < T; t++) {
    for (let j = 0; j < nStates; j++) {
      let bestVal = -Infinity;
      let bestIdx = 0;
      for (let i = 0; i < nStates; i++) {
        const val = delta[t - 1][i] + Math.log(Math.max(A[i][j], 1e-300));
        if (val > bestVal) {
          bestVal = val;
          bestIdx = i;
        }
      }
      delta[t][j] = bestVal + Math.log(Math.max(gaussianPdf(obs[t], means[j], stds[j]), 1e-300));
      psi[t][j] = bestIdx;
    }
  }

  // Backtrace
  const path: number[] = Array(T).fill(0);
  let bestFinal = -Infinity;
  for (let i = 0; i < nStates; i++) {
    if (delta[T - 1][i] > bestFinal) {
      bestFinal = delta[T - 1][i];
      path[T - 1] = i;
    }
  }
  for (let t = T - 2; t >= 0; t--) {
    path[t] = psi[t + 1][path[t + 1]];
  }

  return path;
}

// ---------------------------------------------------------------------------
// Predict: compute posterior state probabilities and n-step forecast
// ---------------------------------------------------------------------------

export function predict(
  observations: number[],
  params: HMMParams,
  forecastHorizon: number,
): HMMPrediction {
  const { nStates, A, means, stds } = params;

  // Forward pass to get current state posteriors
  const { alpha, scales } = forward(observations, params);
  const beta = backward(observations, params, scales);
  const T = observations.length;

  // Compute gamma (posterior state probabilities)
  const gamma: number[][] = Array.from({ length: T }, () => Array(nStates).fill(0));
  for (let t = 0; t < T; t++) {
    let sum = 0;
    for (let i = 0; i < nStates; i++) {
      gamma[t][i] = alpha[t][i] * beta[t][i];
      sum += gamma[t][i];
    }
    if (sum < 1e-300) sum = 1e-300;
    for (let i = 0; i < nStates; i++) gamma[t][i] /= sum;
  }

  const currentStateProbabilities = gamma[T - 1];

  // Most likely state via Viterbi
  const path = viterbi(observations, params);
  const currentState = path[T - 1];

  // n-step forecast: multiply current state distribution by A^n
  const An = matPow(A, forecastHorizon);
  const forecastProbabilities: number[] = Array(nStates).fill(0);
  for (let j = 0; j < nStates; j++) {
    for (let i = 0; i < nStates; i++) {
      forecastProbabilities[j] += currentStateProbabilities[i] * An[i][j];
    }
  }

  // Expected return and volatility (forecast-weighted)
  const expectedReturn = forecastProbabilities.reduce(
    (s, p, i) => s + p * means[i], 0,
  );
  const expectedVolatility = Math.sqrt(
    forecastProbabilities.reduce(
      (s, p, i) => s + p * (stds[i] ** 2 + means[i] ** 2), 0,
    ) - expectedReturn ** 2,
  );

  return {
    currentState,
    stateProbabilities: gamma,
    currentStateProbabilities,
    forecastProbabilities,
    expectedReturn,
    expectedVolatility,
  };
}

// ---------------------------------------------------------------------------
// Matrix power (reused for n-step transition forecast)
// ---------------------------------------------------------------------------

function matPow(M: number[][], n: number): number[][] {
  const size = M.length;
  let result = Array.from({ length: size }, (_, i) =>
    Array.from({ length: size }, (_, j) => (i === j ? 1 : 0)),
  );
  let base = M.map(row => [...row]);

  let power = n;
  while (power > 0) {
    if (power % 2 === 1) {
      result = matMul(result, base);
    }
    base = matMul(base, base);
    power = Math.floor(power / 2);
  }
  return result;
}

function matMul(A: number[][], B: number[][]): number[][] {
  const n = A.length;
  const m = B[0].length;
  const k = B.length;
  const result: number[][] = Array.from({ length: n }, () => Array(m).fill(0));
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < m; j++) {
      for (let l = 0; l < k; l++) {
        result[i][j] += A[i][l] * B[l][j];
      }
    }
  }
  return result;
}

File: research/models/hmm.py
"""Gaussian Hidden Markov Model for financial regime detection.

Wraps hmmlearn.GaussianHMM with a TypeScript-compatible API.

Note on raw-return HMMs:
  Gaussian HMMs maximize likelihood by fitting Gaussian distributions to
  observations. In financial returns, variance differences between regimes
  are typically orders of magnitude larger than mean differences. As a
  result, a 3-state HMM on raw daily returns tends to cluster by
  volatility (e.g. one tight "normal" state + outlier states) rather
  than by return direction. The value of the HMM is in the continuous
  expected_return and expected_volatility forecasts, not in hard state
  labels. For explicit bull/bear/sideways regimes, use the observable
  Markov model in markov.py.
"""

from __future__ import annotations

import math
from dataclasses import dataclass

import numpy as np
from hmmlearn import hmm


# ---------------------------------------------------------------------------
# Types
# ---------------------------------------------------------------------------


@dataclass
class AssetProfile:
    hmm_weight_multiplier: float
    student_t_nu: int
    decay_rate: float


ASSET_PROFILES: dict[str, AssetProfile] = {
    "etf": AssetProfile(hmm_weight_multiplier=1.1, student_t_nu=5, decay_rate=0.97),
    "equity": AssetProfile(hmm_weight_multiplier=0.9, student_t_nu=4, decay_rate=0.96),
    "crypto": AssetProfile(hmm_weight_multiplier=0.5, student_t_nu=3, decay_rate=0.94),
    "commodity": AssetProfile(hmm_weight_multiplier=0.7, student_t_nu=4, decay_rate=0.95),
}


@dataclass
class HMMParams:
    n_states: int
    pi: np.ndarray
    A: np.ndarray
    means: np.ndarray
    stds: np.ndarray


@dataclass
class HMMFitResult:
    params: HMMParams
    log_likelihood: float
    iterations: int
    converged: bool


@dataclass
class HMMPrediction:
    current_state: int
    state_probabilities: np.ndarray
    current_state_probabilities: np.ndarray
    forecast_probabilities: np.ndarray
    expected_return: float
    expected_volatility: float


# ---------------------------------------------------------------------------
# Initialization
# ---------------------------------------------------------------------------


def initialize_hmm(observations: np.ndarray, n_states: int) -> HMMParams:
    """K-means-style quantile initialization with sorted means.

    State 0 = bearish (lowest mean), state N-1 = bullish (highest mean).
    """
    obs = np.asarray(observations).flatten()
    sorted_obs = np.sort(obs)
    n = len(sorted_obs)

    # Quantile-based means
    quantiles = [int(n * i / n_states) for i in range(n_states + 1)]
    means = np.array([
        float(np.mean(sorted_obs[quantiles[i]:quantiles[i + 1]]))
        for i in range(n_states)
    ])

    # Sort means ascending and track permutation
    order = np.argsort(means)
    means = means[order]

    # Per-state stds from quantile slices
    stds = np.array([
        max(float(np.std(sorted_obs[quantiles[i]:quantiles[i + 1]])), 1e-4)
        for i in range(n_states)
    ])[order]

    # Diagonal-dominant transition matrix
    A = np.eye(n_states) * 0.8 + np.ones((n_states, n_states)) * 0.2 / n_states
    A = A / A.sum(axis=1, keepdims=True)

    # Uniform initial distribution
    pi = np.ones(n_states) / n_states

    return HMMParams(n_states=n_states, pi=pi, A=A, means=means, stds=stds)


# ---------------------------------------------------------------------------
# Baum-Welch fitting
# ---------------------------------------------------------------------------


def baum_welch(
    observations: np.ndarray,
    n_states: int = 3,
    max_iterations: int = 100,
    tolerance: float = 1e-4,
    min_std: float = 1e-4,
) -> HMMFitResult:
    """Fit Gaussian HMM via Baum-Welch EM.

    Returns sorted parameters so state 0 = lowest mean, state N-1 = highest.
    Note: on real financial returns the model often collapses to a
    single low-volatility state with a few outlier states. The value is
    in the continuous forecasts (expected_return, expected_volatility),
    not hard regime labels.
    """
    obs = np.asarray(observations).reshape(-1, 1)
    init = initialize_hmm(observations, n_states)

    model = hmm.GaussianHMM(
        n_components=n_states,
        covariance_type="diag",
        n_iter=max_iterations,
        tol=tolerance,
        init_params="",
    )

    model.startprob_ = init.pi.copy()
    model.transmat_ = init.A.copy()
    model.means_ = init.means.reshape(-1, 1).copy()
    model.covars_ = (init.stds ** 2).reshape(-1, 1).copy()

    try:
        model.fit(obs)
        converged = model.monitor_.converged if hasattr(model, "monitor_") else True
        iterations = model.monitor_.iter if hasattr(model, "monitor_") else max_iterations
        log_likelihood = float(model.score(obs))
    except Exception:
        converged = False
        iterations = max_iterations
        log_likelihood = float("-inf")

    # Extract parameters
    means = model.means_.flatten()
    covars = model.covars_.flatten()
    stds = np.sqrt(np.maximum(covars, min_std ** 2))
    A = model.transmat_
    pi = model.startprob_

    # Sort means ascending (bear -> sideways -> bull) and permute everything
    order = np.argsort(means)
    means = means[order]
    stds = stds[order]
    pi = pi[order]
    A = A[np.ix_(order, order)]

    params = HMMParams(n_states=n_states, pi=pi, A=A, means=means, stds=stds)
    return HMMFitResult(
        params=params,
        log_likelihood=log_likelihood,
        iterations=iterations,
        converged=converged,
    )


# ---------------------------------------------------------------------------
# Viterbi decoding
# ---------------------------------------------------------------------------


def viterbi(observations: np.ndarray, params: HMMParams) -> list[int]:
    """Decode most likely state sequence via Viterbi."""
    obs = np.asarray(observations).reshape(-1, 1)
    model = _build_model(params)
    states = model.predict(obs)
    return states.tolist()


# ---------------------------------------------------------------------------
# Prediction
# ---------------------------------------------------------------------------


def predict(
    observations: np.ndarray,
    params: HMMParams,
    forecast_horizon: int,
) -> HMMPrediction:
    """Compute posterior state probabilities and n-step forecast."""
    obs = np.asarray(observations).reshape(-1, 1)
    model = _build_model(params)

    # Posterior probabilities for the last observation
    probs = model.predict_proba(obs)
    current_state_probs = probs[-1]
    current_state = int(np.argmax(current_state_probs))

    # n-step forecast via matrix power
    A_n = mat_pow(params.A, forecast_horizon)
    forecast_probs = A_n[current_state]

    # Expected return and volatility as weighted averages
    expected_return = float(np.dot(forecast_probs, params.means))
    expected_volatility = float(np.dot(forecast_probs, params.stds))

    return HMMPrediction(
        current_state=current_state,
        state_probabilities=current_state_probs,
        current_state_probabilities=current_state_probs,
        forecast_probabilities=forecast_probs,
        expected_return=expected_return,
        expected_volatility=expected_volatility,
    )


# ---------------------------------------------------------------------------
# Matrix helpers
# ---------------------------------------------------------------------------


def mat_pow(A: np.ndarray, n: int) -> np.ndarray:
    """Matrix exponentiation for n-step transition probabilities."""
    if n <= 0:
        return np.eye(A.shape[0])
    return np.linalg.matrix_power(A, n)


# ---------------------------------------------------------------------------
# Volatility HMM helper
# ---------------------------------------------------------------------------


def fit_2state_return_hmm(
    returns: np.ndarray,
    max_iterations: int = 50,
    tolerance: float = 1e-3,
) -> dict:
    """Fit a 2-state Gaussian HMM on raw returns.

    Labels states by volatility (low-vol = "calm", high-vol = "volatile").
    Returns stationary probabilities, per-state means/vols, and current
    state information. More balanced than 3-state on volatile assets.

    Returns
    -------
    dict
        converged, state_labels, state_probs, state_means, state_vols,
        expected_return, expected_volatility, current_state, current_state_prob.
    """
    arr = np.asarray(returns)
    if len(arr) < 20:
        return {
            "converged": False,
            "state_labels": ["calm", "volatile"],
            "state_probs": [0.5, 0.5],
            "state_means": [0.0, 0.0],
            "state_vols": [0.01, 0.02],
            "expected_return": 0.0,
            "expected_volatility": 0.01,
            "current_state": 0,
            "current_state_prob": 0.5,
        }

    result = baum_welch(
        arr,
        n_states=2,
        max_iterations=max_iterations,
        tolerance=tolerance,
    )

    if not result.converged:
        return {
            "converged": False,
            "state_labels": ["calm", "volatile"],
            "state_probs": [0.5, 0.5],
            "state_means": [0.0, 0.0],
            "state_vols": [0.01, 0.02],
            "expected_return": 0.0,
            "expected_volatility": 0.01,
            "current_state": 0,
            "current_state_prob": 0.5,
        }

    params = result.params
    # Sort by volatility (ascending): state 0 = calm, state 1 = volatile
    order = np.argsort(params.stds)
    means = params.means[order]
    vols = params.stds[order]
    A = params.A[np.ix_(order, order)]

    # Stationary distribution via power iteration (more stable than eig)
    stationary = np.ones(2) / 2
    for _ in range(100):
        next_stationary = stationary @ A
        if np.allclose(next_stationary, stationary, atol=1e-10):
            break
        stationary = next_stationary
    stationary = np.maximum(stationary, 0.0)
    stationary = stationary / stationary.sum()

    # Current state probabilities from last observation
    model = _build_model(
        HMMParams(n_states=2, pi=params.pi[order], A=A, means=means, stds=vols)
    )
    obs = arr.reshape(-1, 1)
    probs = model.predict_proba(obs)
    current_state_probs = probs[-1]
    current_state = int(np.argmax(current_state_probs))

    expected_return = float(np.dot(stationary, means))
    expected_volatility = float(np.dot(stationary, vols))

    return {
        "converged": True,
        "state_labels": ["calm", "volatile"],
        "state_probs": stationary.tolist(),
        "state_means": means.tolist(),
        "state_vols": vols.tolist(),
        "expected_return": expected_return,
        "expected_volatility": expected_volatility,
        "current_state": current_state,
        "current_state_prob": float(current_state_probs[current_state]),
    }


def fit_volatility_hmm(
    returns: np.ndarray,
    vol_window: int = 5,
    n_states: int = 2,
    max_iterations: int = 50,
    tolerance: float = 1e-3,
) -> float:
    """Fit 2-state HMM on rolling realised volatility.

    Returns a vol scale factor clamped to [0.5, 2.0].
    """
    arr = np.asarray(returns)
    # Rolling realised volatility (std of returns over window)
    vols = []
    for i in range(vol_window - 1, len(arr)):
        window = arr[i - vol_window + 1:i + 1]
        vols.append(float(np.std(window, ddof=1)))
    vols = np.array(vols)

    if len(vols) < n_states * 10:
        return 1.0

    result = baum_welch(
        vols,
        n_states=n_states,
        max_iterations=max_iterations,
        tolerance=tolerance,
    )
    if not result.converged:
        return 1.0

    # Identify high-vol state
    params = result.params
    high_vol_state = int(np.argmax(params.means))
    low_vol_state = int(np.argmin(params.means))

    # Current vol is the last observation
    current_vol = float(vols[-1])

    # Baseline: midpoint between high and low state means
    baseline = (params.means[high_vol_state] + params.means[low_vol_state]) / 2.0
    if baseline <= 0:
        return 1.0

    scale = current_vol / baseline
    return float(np.clip(scale, 0.5, 2.0))


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------


def _build_model(params: HMMParams) -> hmm.GaussianHMM:
    """Reconstruct a fitted GaussianHMM from parameters."""
    model = hmm.GaussianHMM(
        n_components=params.n_states,
        covariance_type="diag",
        n_iter=1,
        init_params="",
    )
    model.startprob_ = params.pi.copy()
    model.transmat_ = params.A.copy()
    model.means_ = params.means.reshape(-1, 1).copy()
    model.covars_ = (params.stds ** 2).reshape(-1, 1).copy()
    return model
